### PR TITLE
ROCm: Fix AITER attention for Qwen3-Coder-Next hybrid models

### PR DESCRIPTION
**Pull request by jhinpan**
## Goal

- Fix v_head_dim init for hybrid models (layer0 can be GDN, not full attention).
- Add probe-based CK batch_prefill support detection and triton extend fallback for head_dim>128.
- Guard CUDA-only CUTLASS/CuTe-DSL imports and keep ROCm on triton causal_conv1d.

## Motivation

Qwen3-Coder-Next is a hybrid model (GDN + attention). On ROCm with AITER attention backend enabled, we hit runtime issues because:

1. `v_head_dim` was inferred from `token_to_kv_pool.get_value_buffer(0)` which can point to a non-attention layer buffer when layer0 is GDN.
2. CK `mha_batch_prefill` availability for `head_dim > 128` depends on the local AITER build. When unsupported, prefill can fail (or be unstable).
3. `hybrid_linear_attn_backend.py` imported CUDA-only CUTLASS/CuTe-DSL bits unconditionally in the non-CPU path, which breaks ROCm imports.

This PR makes Qwen3-Coder-Next runnable and benchmarkable across both `--attention-backend triton` and `--attention-backend aiter` on ROCm.

## Modifications

- `python/sglang/srt/layers/attention/aiter_backend.py`
  - Fix `v_head_dim` initialization by using `model_config.v_head_dim` instead of `token_to_kv_pool.get_value_buffer(0)`.
  - Add a small probe at backend init to detect whether CK batch-prefill supports the current head shape (non-MLA, `head_dim > 128`).
    - If probe fails, fall back to triton `extend_attention_fwd` in `forward_extend`.
  - Skip the probe when `skip_prefill=True` (decode-only backends like speculative draft backends).

- `python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py`
  - Guard CUDA-only `cutedsl_gdn` import behind `is_cuda()`.
  - On ROCm/HIP, keep using the triton `causal_conv1d_triton` implementation (do not import/override with CUDA-only implementations).
  - Only enable CuTe-DSL GDN decode when running on CUDA.

### GSM8K on Qwen3-Coder-Next (TP=8)

Rerun: 2026-02-11 (GSM8K test set has 1319 questions; `--num-questions 2000` runs all of them)

- Triton attention backend:
  - Accuracy: **0.947**
  - Invalid: **0.000**
  - Latency: **33.062 s**
  - Output throughput: **5198.031 tok/s**

- AITER attention backend:
  - Accuracy: **0.947**
  - Invalid: **0.000**
  - Latency: **32.303 s**
  - Output throughput: **5264.421 tok/s**

Summary:

| Backend | Accuracy | Invalid | Latency | Output tok/s |
| --- | --- | --- | --- | --- |
| triton | 0.947 | 0.000 | 33.062 s | 5198.031 |
| aiter | 0.947 | 0.000 | 32.303 s | 5264.421 |

Commands:

```bash
# Triton
MODEL_PATH=/path/to/Qwen3-Coder-Next
SGLANG_USE_AITER=1 python3 -m sglang.launch_server \
  --model-path "$MODEL_PATH" \
  --tp 8 --trust-remote-code \
  --attention-backend triton \
  --host 0.0.0.0 --port 30000

PYTHONPATH=python python3 benchmark/gsm8k/bench_sglang.py \
  --num-questions 2000 --parallel 512 --port 30000 --backend srt

# AITER
SGLANG_USE_AITER=1 python3 -m sglang.launch_server \
  --model-path "$MODEL_PATH" \
  --tp 8 --trust-remote-code \
  --attention-backend aiter \
  --host 0.0.0.0 --port 30000

PYTHONPATH=python python3 benchmark/gsm8k/bench_sglang.py \
  --num-questions 2000 --parallel 512 --port 30000 --backend srt
```

### Qwen3-Coder-Next Serving Benchmark Matrix on MI355X (TP=1/4/8)

Rerun: 2026-02-11

Parameters: random dataset, 50 prompts, input len 512, output len 256, request rate 5 req/s.

| Metric | Triton TP=1 | Triton TP=4 | Triton TP=8 | AITER TP=1 | AITER TP=4 | AITER TP=8 |
| --- | --- | --- | --- | --- | --- | --- |
| Status | OK | OK | OK | OK | OK | OK |
| Successful requests | 50/50 | 50/50 | 50/50 | 50/50 | 50/50 | 50/50 |
| Request throughput (req/s) | 3.85 | 4.08 | 4.10 | 3.84 | 4.06 | 4.08 |
| Input tok/s | 1,090 | 1,154 | 1,160 | 1,088 | 1,150 | 1,156 |
| Output tok/s | 438 | 464 | 466 | 437 | 462 | 464 |
| Peak output tok/s | 686 | 752 | 753 | 685 | 773 | 760 |
| Total tok/s | 1,528 | 1,617 | 1,626 | 1,525 | 1,611 | 1,621 |
| Concurrency | 13.60 | 10.00 | 9.65 | 13.44 | 10.30 | 10.02 |
| Mean E2E Latency (ms) | 3,533 | 2,453 | 2,354 | 3,496 | 2,538 | 2,453 |
| Median TTFT (ms) | 86.7 | 82.3 | 80.1 | 83.7 | 84.0 | 83.3 |
| Median TPOT (ms) | 29.8 | 21.1 | 20.4 | 29.8 | 21.7 | 21.2 |

Command:

```bash
MODEL_PATH=/path/to/Qwen3-Coder-Next
PORT=30000

# Server
python3 -m sglang.launch_server \
  --model-path "$MODEL_PATH" \
  --tp {1|4|8} \
  --trust-remote-code \
  --attention-backend {triton|aiter} \
  --port "$PORT"

# Benchmark
python3 -m sglang.bench_serving \
  --backend sglang \
  --host 127.0.0.1 \
  --port "$PORT" \
  --model "$MODEL_PATH" \
  --dataset-name random \
  --random-input-len 512 \
  --random-output-len 256 \
  --num-prompts 50 \
  --request-rate 5
```

## Checklist

- [x] Format your code according to the Format code with pre-commit.
- [ ] Add unit tests according to the Run and add unit tests.
- [ ] Update documentation according to Write documentations.
- [x] Provide accuracy and speed benchmark results.
- [x] Follow the SGLang code style guidance.

## Review Process

1. Ping Merge Oncalls to start the PR flow.
2. Get approvals from CODEOWNERS and other reviewers.
3. Trigger CI tests with comments.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/layers/attention/aiter_backend.py [hunk 1]
```diff
@@ -123,6 +123,57 @@ def __init__(
...
+                    causal=True,
+                )
+            except RuntimeError:
+                self._needs_triton_fallback = True
+                logger.warning(
+                    "CK batch_prefill does not support head_dim=%d, "
+                    "falling back to triton extend kernel",
+                    self.head_dim,
+                )
+
```
Comment (sunxxuns) [inline python/sglang/srt/layers/attention/aiter_backend.py:185]: i feel this should be wrapped into a function and put it somewhere with other ck related funcs.

File: python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py [hunk 2]
```diff
@@ -33,25 +33,33 @@
...
     from sglang.srt.layers.attention.mamba.causal_conv1d import (
         causal_conv1d_fn as causal_conv1d_fn_cuda,
     )
 
     causal_conv1d_fn = causal_conv1d_fn_cuda
+elif is_hip():
+    # On ROCm/HIP, sgl_kernel.causal_conv1d_fwd is not available.
+    # Use the triton-based causal_conv1d_fn already imported above from
+    # causal_conv1d_triton (lines 17-21) — no override needed.
+    pass
```
Comment (sunxxuns) [inline python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py]: do need this

File: python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py [hunk 3]
```diff
@@ -33,25 +33,33 @@
...
         CHUNK_SIZE as FLA_CHUNK_SIZE,
     )
     from sglang.srt.layers.attention.fla.kda import chunk_kda
 
+if is_cuda():
+    # cutedsl_gdn requires NVIDIA CUTLASS (not available on ROCm/HIP)
+    from sglang.jit_kernel.cutedsl_gdn import (
+        cutedsl_fused_sigmoid_gating_delta_rule_update,
+    )
+
```
Comment (sunxxuns) [inline python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py:51]: why do we need this for hip? if this is to be skipped if is-hip is true, changed it in the original line

File: python/sglang/srt/layers/attention/aiter_backend.py [hunk 4]
```diff
@@ -1303,23 +1354,50 @@ def forward_extend(
...
+                    self.forward_metadata.max_kv_len,
+                    causal=True,
+                    logits_soft_cap=self.logits_soft_cap,
+                    alibi_slopes=None,
+                    return_lse=False,
+                    return_attn_probs=False,
+                )
 
-            return o.view(-1, layer.tp_q_head_num * layer.head_dim)
+                return o.view(-1, layer.tp_q_head_num * layer.head_dim)
```
Comment (sunxxuns) [inline python/sglang/srt/layers/attention/aiter_backend.py]: avoid branching at high level

File: python/sglang/srt/layers/attention/aiter_backend.py (hunk 1 above)
Comment (jhinpan) [reply python/sglang/srt/layers/attention/aiter_backend.py:185]: Good suggestion. I refactored the CK batch_prefill probe into a small helper (`_ck_batch_prefill_is_supported`) colocated in `aiter_backend.py`, and `__init__` now calls it once to set `_use_ck_batch_prefill`. In `forward_extend` I switched to an early-return triton fallback when unsupported, so we dont duplicate / rewrite the whole CK branch.

File: python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py (hunk 2 above)
Comment (jhinpan) [reply python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py]: Agreed. This `is_hip` import/branch wasn\x27t needed. I removed it and now ROCm/HIP just keeps the default triton `causal_conv1d_triton` path (CUDA still overrides to the CUDA kernel).

File: python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py (hunk 3 above)
Comment (jhinpan) [reply python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py:51]: Yep. I removed the HIP special-casing and left the original line intact: on ROCm we simply do not override `causal_conv1d_fn`, so it stays on the triton implementation. This keeps the intent (CUDA uses CUDA kernel; others use triton) without extra branches.
